Goals: 
- Deal with null data points   
- Clean the data (getting rid of the units)
- Data split (train / test)
- Save the clean data in data/processed 

In [532]:
import pandas as pd

In [533]:
file_loc = "../data/raw/Car details v3.csv"
df = pd.read_csv(file_loc)
df.notna().sum()

name             8128
year             8128
selling_price    8128
km_driven        8128
fuel             8128
seller_type      8128
transmission     8128
owner            8128
mileage          7907
engine           7907
max_power        7913
torque           7906
seats            7907
dtype: int64

In [534]:
8128-7906 #=222
(8128-7906)/8128*100

2.731299212598425

Torque has the least number of Non-none values. If we were to do dropNA, we would lose at least $8128-7906 = 222$ values. This will correspond to $222/8128*100 = 2.73\%$ data loss.  

In [535]:
df = df.dropna()
#df.notna().sum()
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7906 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           7906 non-null   object 
 1   year           7906 non-null   int64  
 2   selling_price  7906 non-null   int64  
 3   km_driven      7906 non-null   int64  
 4   fuel           7906 non-null   object 
 5   seller_type    7906 non-null   object 
 6   transmission   7906 non-null   object 
 7   owner          7906 non-null   object 
 8   mileage        7906 non-null   object 
 9   engine         7906 non-null   object 
 10  max_power      7906 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7906 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 864.7+ KB


In the dataframe, NA values are overlapping. That is, many rows have NA column entries. We will drop those rows.

In [536]:
df.sample(5)

name  year  selling_price  km_driven  \
1217                          Tata Nano CX  2013          80000      30000   
7279  Volkswagen Polo 2015-2019 GT 1.5 TDI  2016         591000      60000   
5497              Maruti Wagon R VXI BSIII  2012         260000      31000   
2406               Hyundai i10 Sportz 1.1L  2010         200000      70000   
847                Hyundai i10 Sportz 1.1L  2016         425000      15000   

        fuel seller_type transmission         owner     mileage   engine  \
1217  Petrol  Individual       Manual   First Owner   25.4 kmpl   624 CC   
7279  Diesel      Dealer       Manual   First Owner  21.49 kmpl  1498 CC   
5497  Petrol  Individual       Manual   First Owner   18.9 kmpl   998 CC   
2406  Petrol  Individual       Manual  Second Owner  19.81 kmpl  1086 CC   
847   Petrol  Individual       Manual   First Owner  19.81 kmpl  1086 CC   

      max_power               torque  seats  
1217  37.48 bhp  51Nm@ 4000+/-500rpm    4.0  
7279  108.5 bhp  250nm@ 1500-3000rpm    5.0  
5497   67.1 bhp        90Nm@ 3500rpm    5.0  
2406  68.05 bhp     99.04Nm@ 4500rpm    5.0  
847   68.05 bhp     99.04Nm@ 4500rpm    5.0

The columns "selling_price", "km_driven", "year", and "seats" have float values. Therefore, no need to change those. The columns "mileage", "engine", and "max_power" have units. We may need to do unit conversion and coversion to float values. The "torque" column is weird, need to pay special attention. The columns "fuel", "seller_type", "transmission", and "owner" are categorical features.

First, we will work on the "engine" column

In [537]:
df["engine"].apply(lambda string : string.split(" ")[1]).unique()

array(['CC'], dtype=object)

The "engine" column has only "CC" (cubic centimeters) unit. We only need to grab the corresponding value.

In [538]:
df["engine_value_cc"] = df["engine"].apply(lambda string : float(string.split(" ")[0]))
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7906 entries, 0 to 8127
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             7906 non-null   object 
 1   year             7906 non-null   int64  
 2   selling_price    7906 non-null   int64  
 3   km_driven        7906 non-null   int64  
 4   fuel             7906 non-null   object 
 5   seller_type      7906 non-null   object 
 6   transmission     7906 non-null   object 
 7   owner            7906 non-null   object 
 8   mileage          7906 non-null   object 
 9   engine           7906 non-null   object 
 10  max_power        7906 non-null   object 
 11  torque           7906 non-null   object 
 12  seats            7906 non-null   float64
 13  engine_value_cc  7906 non-null   float64
dtypes: float64(2), int64(3), object(9)
memory usage: 926.5+ KB


Do the same for max power.

In [539]:
df["max_power"].apply(lambda string: string.split(" ")[1]).unique()

array(['bhp'], dtype=object)

In [540]:
df["max_power_bph"] = df["max_power"].apply(lambda string: float(string.split(" ")[0]))
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7906 entries, 0 to 8127
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             7906 non-null   object 
 1   year             7906 non-null   int64  
 2   selling_price    7906 non-null   int64  
 3   km_driven        7906 non-null   int64  
 4   fuel             7906 non-null   object 
 5   seller_type      7906 non-null   object 
 6   transmission     7906 non-null   object 
 7   owner            7906 non-null   object 
 8   mileage          7906 non-null   object 
 9   engine           7906 non-null   object 
 10  max_power        7906 non-null   object 
 11  torque           7906 non-null   object 
 12  seats            7906 non-null   float64
 13  engine_value_cc  7906 non-null   float64
 14  max_power_bph    7906 non-null   float64
dtypes: float64(3), int64(3), object(9)
memory usage: 988.2+ KB


Now, working on the "mileage" column.

In [541]:
df["mileage"].apply(lambda string: string.split(" ")[1]).unique()

array(['kmpl', 'km/kg'], dtype=object)

We want to convert to a common unit. We will convert km/kg to kmpl. We need to check which fuel uses kmpl and which used km/kg. Then, we will multiply by density of the fluid. $$\frac{km}{L} = \frac{km}{kg} * \frac{kg}{L}. $$

In [542]:
fuel_types = df["fuel"].unique() #['Diesel' 'Petrol' 'LPG' 'CNG']

for entry in fuel_types:
    fuel_units = df[df["fuel"] == entry]["mileage"].apply(lambda string : string.split(" ")[1]).unique()
    print(f"{entry:<20}{fuel_units}")

Diesel              ['kmpl']
Petrol              ['kmpl']
LPG                 ['km/kg']
CNG                 ['km/kg']


We need to find the density of LPG and CNG.

In [543]:
density_LPG = 0.4*1.96 + 0.6 * 1.724 # Kg/L from https://www.rippleinfra.com/lpg.html  Taking 4:6 ratio of propane : butane 

density_CNG = 200/1000 # check both values. 200 is kg/m^3

In [544]:
#making categorical variables for fuel types.

for entry in fuel_types:
    df[f"fuel_{entry}"] = df["fuel"].apply(lambda string : 1 if string == entry else 0)

df.sample(5)

name  year  selling_price  \
461       Toyota Innova 2.5 G (Diesel) 7 Seater  2012         730000   
6866           Mahindra TUV 300 mHAWK100 T8 AMT  2016         750000   
7415  Hyundai EON 1.0 Kappa Magna Plus Optional  2018         350000   
2250                Maruti Ritz VXi (ABS) BS IV  2010         170000   
612                         Maruti Alto 800 LXI  2013         200000   

      km_driven    fuel seller_type transmission         owner     mileage  \
461      250000  Diesel  Individual       Manual   First Owner  12.99 kmpl   
6866     110000  Diesel  Individual    Automatic   First Owner  18.49 kmpl   
7415      10000  Petrol  Individual       Manual  Second Owner   20.3 kmpl   
2250     120000  Petrol  Individual       Manual  Second Owner   18.5 kmpl   
612       24000  Petrol  Individual       Manual   First Owner  22.74 kmpl   

       engine  max_power               torque  seats  engine_value_cc  \
461   2494 CC    100 bhp  200Nm@ 1400-3400rpm    7.0           2494.0   
6866  1493 CC    100 bhp  240Nm@ 1600-2800rpm    7.0           1493.0   
7415   998 CC  68.05 bhp     94.14Nm@ 3500rpm    5.0            998.0   
2250  1197 CC  85.80 bhp       114Nm@ 4000rpm    5.0           1197.0   
612    796 CC   47.3 bhp        69Nm@ 3500rpm    5.0            796.0   

      max_power_bph  fuel_Diesel  fuel_Petrol  fuel_LPG  fuel_CNG  
461          100.00            1            0         0         0  
6866         100.00            1            0         0         0  
7415          68.05            0            1         0         0  
2250          85.80            0            1         0         0  
612           47.30            0            1         0         0

We will make a column called "mileage_conversion_factor". This column contain the conversion factor for each type of fuel. For example, for petrol we already have mileage in kmpl, therefore the conversion factor is 1. For CNG, it will be the density.

In [545]:
df["mileage_coversion_factor"] = df["fuel_Diesel"] + df["fuel_Petrol"] + density_CNG * df["fuel_CNG"] + density_LPG * df["fuel_LPG"] # multiply by this number to convert

Now convert all mileage values to kmpl.

In [546]:
df["mileage_kmpl"] = df["mileage"].apply(lambda string: float(string.split(" ")[0])) * df["mileage_coversion_factor"]
df.head(15)

name  year  selling_price  km_driven  \
0                 Maruti Swift Dzire VDI  2014         450000     145500   
1           Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2               Honda City 2017-2020 EXi  2006         158000     140000   
3              Hyundai i20 Sportz Diesel  2010         225000     127000   
4                 Maruti Swift VXI BSIII  2007         130000     120000   
5          Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
6           Maruti Wagon R LXI DUO BSIII  2007          96000     175000   
7                     Maruti 800 DX BSII  2001          45000       5000   
8                       Toyota Etios VXD  2011         350000      90000   
9   Ford Figo Diesel Celebration Edition  2013         200000     169000   
10       Renault Duster 110PS Diesel RxL  2014         500000      68000   
11                         Maruti Zen LX  2005          92000     100000   
12                Maruti Swift Dzire VDi  2009         280000     140000   
14              Maruti Wagon R LXI Minor  2009         180000      90000   
15  Mahindra KUV 100 mFALCON G80 K8 5str  2016         400000      40000   

      fuel seller_type transmission         owner     mileage   engine  ...  \
0   Diesel  Individual       Manual   First Owner   23.4 kmpl  1248 CC  ...   
1   Diesel  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  ...   
2   Petrol  Individual       Manual   Third Owner   17.7 kmpl  1497 CC  ...   
3   Diesel  Individual       Manual   First Owner   23.0 kmpl  1396 CC  ...   
4   Petrol  Individual       Manual   First Owner   16.1 kmpl  1298 CC  ...   
5   Petrol  Individual       Manual   First Owner  20.14 kmpl  1197 CC  ...   
6      LPG  Individual       Manual   First Owner  17.3 km/kg  1061 CC  ...   
7   Petrol  Individual       Manual  Second Owner   16.1 kmpl   796 CC  ...   
8   Diesel  Individual       Manual   First Owner  23.59 kmpl  1364 CC  ...   
9   Diesel  Individual       Manual   First Owner   20.0 kmpl  1399 CC  ...   
10  Diesel  Individual       Manual  Second Owner  19.01 kmpl  1461 CC  ...   
11  Petrol  Individual       Manual  Second Owner   17.3 kmpl   993 CC  ...   
12  Diesel  Individual       Manual  Second Owner   19.3 kmpl  1248 CC  ...   
14  Petrol  Individual       Manual  Second Owner   18.9 kmpl  1061 CC  ...   
15  Petrol  Individual       Manual   First Owner  18.15 kmpl  1198 CC  ...   

                      torque seats  engine_value_cc  max_power_bph  \
0             190Nm@ 2000rpm   5.0           1248.0          74.00   
1        250Nm@ 1500-2500rpm   5.0           1498.0         103.52   
2      12.7@ 2,700(kgm@ rpm)   5.0           1497.0          78.00   
3   22.4 kgm at 1750-2750rpm   5.0           1396.0          90.00   
4      11.5@ 4,500(kgm@ rpm)   5.0           1298.0          88.20   
5          113.75nm@ 4000rpm   5.0           1197.0          81.86   
6       7.8@ 4,500(kgm@ rpm)   5.0           1061.0          57.50   
7              59Nm@ 2500rpm   4.0            796.0          37.00   
8        170Nm@ 1800-2400rpm   5.0           1364.0          67.10   
9             160Nm@ 2000rpm   5.0           1399.0          68.10   
10            248Nm@ 2250rpm   5.0           1461.0         108.45   
11             78Nm@ 4500rpm   5.0            993.0          60.00   
12            190Nm@ 2000rpm   5.0           1248.0          73.90   
14             84Nm@ 3500rpm   5.0           1061.0          67.00   
15       115Nm@ 3500-3600rpm   5.0           1198.0          82.00   

    fuel_Diesel  fuel_Petrol  fuel_LPG  fuel_CNG  mileage_coversion_factor  \
0             1            0         0         0                    1.0000   
1             1            0         0         0                    1.0000   
2             0            1         0         0                    1.0000   
3             1            0         0         0                    1.0000   
4             0            1         0     

Next, we will extract data from the torque values. Unfortunately, the torque values have different types of units. From the first glance, there is Nm (Newton-meter) and kgm (kg-meter) for the torque. Also, there are rpm vales. We would like to know the kind of units we are given.

The following function takes a string and returns the same string without any numbers.

In [547]:
def eat_numbers(text:str ) -> str:
    """Takes a string and outputs the same string without numbers. Example: "No 1 DJ" -> "No  DJ", "50kg" -> "kg" """
    return ''.join(["" if char.isdigit() else char for char in text])

assert(eat_numbers("No 1 DJ") == "No  DJ")
assert(eat_numbers("115Nm@ 3500-3600rpm") == "Nm@ -rpm")

import re 

def replace_numbers_with_star(text : str) -> str:
    """Takes a string replaces a number sequence with *. Example: "No 10 DJ" -> "No * DJ", "50kg over 500m" -> "*kg over *m" """
    new_text =  re.sub(r'\d+', '*', text)
    return new_text.replace('*.*', '*').replace(' ', '').lower()

replace_numbers_with_star("50kg over 500m")

'*kgover*m'

Let's check what units are there.

In [548]:
df["torque"].apply(replace_numbers_with_star).unique()

array(['*nm@*rpm', '*nm@*-*rpm', '*@*,*(kgm@rpm)', '*kgmat*-*rpm',
       '*kgm@*rpm', '*nm@*~*rpm', '*nmat*rpm', '*@*-*rpm', '*nm',
       '*kgmat*rpm', '*kgmat*,*rpm', '*@*-*(kgm@rpm)', '*nmat*-*rpm',
       '*@*,*-*,*(kgm@rpm)', '*nm(*kgm)@*rpm', '*nm@*-*', '*nm@*+/-*rpm',
       '*@*,*+/-*(nm@rpm)', '*@*-*', '*(*)@*', '*nm/*rpm', '*@*(kgm@rpm)',
       '*nm@*,*rpm', '*nm@*', '*/*', '*nmat*,*-*,*rpm'], dtype=object)

In [549]:
df = df.dropna()
#df.info()
columns = list(df.columns)
categorical_features = ['fuel', 'seller_type', 'transmission', 'owner', 'seats']

for feature in categorical_features:
    print(f"{feature:<20} {df[feature].unique()}")

fuel                 ['Diesel' 'Petrol' 'LPG' 'CNG']
seller_type          ['Individual' 'Dealer' 'Trustmark Dealer']
transmission         ['Manual' 'Automatic']
owner                ['First Owner' 'Second Owner' 'Third Owner' 'Fourth & Above Owner'
 'Test Drive Car']
seats                [ 5.  4.  7.  8.  6.  9. 10. 14.  2.]


In [550]:
df[df["fuel"]=="CNG"].head(2)

name  year  selling_price  km_driven fuel  \
35   Maruti Alto 800 CNG LXI Optional  2019         330000      10000  CNG   
203            Maruti Wagon R LXI CNG  2016         360000      50000  CNG   

    seller_type transmission         owner      mileage  engine  ...  \
35   Individual       Manual  Second Owner  33.44 km/kg  796 CC  ...   
203  Individual       Manual   First Owner   26.6 km/kg  998 CC  ...   

            torque seats  engine_value_cc  max_power_bph  fuel_Diesel  \
35   60Nm@ 3500rpm   4.0            796.0          40.30            0   
203  77Nm@ 3500rpm   5.0            998.0          58.16            0   

     fuel_Petrol  fuel_LPG  fuel_CNG  mileage_coversion_factor  mileage_kmpl  
35             0         0         1                       0.2         6.688  
203            0         0         1                       0.2         5.320  

[2 rows x 21 columns]

In [551]:
df["fuel"].unique()

array(['Diesel', 'Petrol', 'LPG', 'CNG'], dtype=object)

In [552]:
df["fuel_Diesel"] = pd.get_dummies(df["fuel"])["Diesel"]

In [553]:
for feature in categorical_features:
    for category in df[feature].unique():
        df[f"{feature}_{category}"] = pd.get_dummies(df[feature], dtype=int)[category]

df.head(15)

df.columns

Index(['name', 'year', 'selling_price', 'km_driven', 'fuel', 'seller_type',
       'transmission', 'owner', 'mileage', 'engine', 'max_power', 'torque',
       'seats', 'engine_value_cc', 'max_power_bph', 'fuel_Diesel',
       'fuel_Petrol', 'fuel_LPG', 'fuel_CNG', 'mileage_coversion_factor',
       'mileage_kmpl', 'seller_type_Individual', 'seller_type_Dealer',
       'seller_type_Trustmark Dealer', 'transmission_Manual',
       'transmission_Automatic', 'owner_First Owner', 'owner_Second Owner',
       'owner_Third Owner', 'owner_Fourth & Above Owner',
       'owner_Test Drive Car', 'seats_5.0', 'seats_4.0', 'seats_7.0',
       'seats_8.0', 'seats_6.0', 'seats_9.0', 'seats_10.0', 'seats_14.0',
       'seats_2.0'],
      dtype='object')

In [554]:
df["engine"].apply(lambda string : string.split(" ")[1]).unique()

array(['CC'], dtype=object)

In [555]:
df["engine_value_cc"] = df["engine"].apply(lambda string : float(string.split(" ")[0]))

In [556]:
df["max_power"].apply(lambda string: string.split(" ")[1]).unique()

array(['bhp'], dtype=object)

In [557]:
df["max_power_bph"] = df["max_power"].apply(lambda string: float(string.split(" ")[0]))

In [558]:
df["mileage"].apply(lambda string : string.split(" ")[1]).unique()

array(['kmpl', 'km/kg'], dtype=object)

In [560]:
df[df.duplicated()==True]


name  year  selling_price  \
291                   Hyundai Grand i10 Sportz  2017         450000   
296                           Maruti Swift VXI  2012         330000   
370   Jaguar XE 2016-2019 2.0L Diesel Prestige  2017        2625000   
371                              Lexus ES 300h  2019        5150000   
372             Jaguar XF 2.0 Diesel Portfolio  2017        3200000   
...                                        ...   ...            ...   
7987             Renault Captur 1.5 Diesel RXT  2018        1265000   
7988                  Maruti Ciaz Alpha Diesel  2019        1025000   
8117                    Maruti Swift Dzire VDI  2015         625000   
8126                           Tata Indigo CR4  2013         290000   
8127                           Tata Indigo CR4  2013         290000   

      km_driven    fuel seller_type transmission         owner     mileage  \
291       35000  Petrol  Individual       Manual   First Owner   18.9 kmpl   
296       50000  Petrol  Individual       Manual  Second Owner   18.6 kmpl   
370        9000  Diesel      Dealer    Automatic   First Owner   13.6 kmpl   
371       20000  Petrol      Dealer    Automatic   First Owner  22.37 kmpl   
372       45000  Diesel      Dealer    Automatic   First Owner  19.33 kmpl   
...         ...     ...         ...          ...           ...         ...   
7987      12000  Diesel  Individual       Manual   First Owner  20.37 kmpl   
7988      32000  Diesel  Individual       Manual   First Owner  28.09 kmpl   
8117      50000  Diesel  Individual       Manual   First Owner  26.59 kmpl   
8126      25000  Diesel  Individual       Manual   First Owner  23.57 kmpl   
8127      25000  Diesel  Individual       Manual   First Owner  23.57 kmpl   

       engine  ... owner_Test Drive Car seats_5.0  seats_4.0  seats_7.0  \
291   1197 CC  ...                    0         1          0          0   
296   1197 CC  ...                    0         1          0          0   
370   1999 CC  ...                    0         1          0          0   
371   2487 CC  ...                    0         1          0          0   
372   1999 CC  ...                    0         1          0          0   
...       ...  ...                  ...       ...        ...        ...   
7987  1461 CC  ...                    0         1          0          0   
7988  1248 CC  ...                    0         1          0          0   
8117  1248 CC  ...                    0         1          0          0   
8126  1396 CC  ...                    0         1          0          0   
8127  1396 CC  ...                    0         1          0          0   

      seats_8.0  seats_6.0  seats_9.0  seats_10.0  seats_14.0  seats_2.0  
291           0          0          0           0           0          0  
296           0          0          0           0           0          0  
370           0          0          0           0           0          0  
371           0          0          0           0           0          0  
372           0          0          0           0           0          0  
...         ...        ...        ...         ...         ...        ...  
7987          0          0          0           0           0          0  
7988          0          0          0           0           0          0  
8117          0          0          0           0           0          0  
8126          0          0          0           0           0          0  
8127          0          0          0           0           0          0  

[1189 rows x 40 columns]